azurelinuxagent/ga/utils/exthandler_utils.py

```diff
@@ -0,0 +1,343 @@
+from azurelinuxagent.common.exception import ExtensionHandlerConfigurationError
+from azurelinuxagent.common.future import ustr
+from azurelinuxagent.common.osutil import get_osutil
+from azurelinuxagent.common.version import AGENT_NAME
+
+MEMORY_OOM_KILL_OPTIONS = ["enabled", "disabled"]
+DEFAULT_CORES_COUNT = -1
+
+AGENT_CGROUP_NAME = "WALinuxAgent"
+DEFAULT_CPU_LIMIT_AGENT = 10.0
+DEFAULT_CPU_LIMIT_EXT = 40.0
+DEFAULT_MEM_LIMIT_MIN_MB_FOR_EXTN = 256.0  # mb, applies to agent and extensions
+DEFAULT_MEM_LIMIT_MAX_MB_FOR_AGENT = 512.0  # mb, applies to agent only
+DEFAULT_MEM_LIMIT_PCT_FOR_EXTN = 15.0  # percent, applies to extensions
+
+CGROUPS_ENFORCE_DEFAULT_LIMITS = False  # Currently, we don't want to enforce defaults as well.
+
+
+class HandlerConfiguration(object):
+    @staticmethod
+    def send_handler_configuration_event(name=AGENT_NAME, message="", operation=WALAEventOperation.Unknown,
+                                         is_success=True, log_event=True):
+        if message:
+            add_event(name=name, message=message, op=operation, is_success=is_success, log_event=log_event)
+
+    def __init__(self, handler_config_json, name=None):
+        self._data = handler_config_json
+        self._name = name
+        self.resource_config = None
+        self.resource_limits = None
+
+        try:
+            if handler_config_json is None:
+                raise ExtensionHandlerConfigurationError('Malformed handler configuration file. '
+                                                  'No data present')
+            if "handlerConfiguration" not in handler_config_json:
+                raise ExtensionHandlerConfigurationError('Malformed handler configuration file. '
+                                                  '"HandlerConfiguration" key not present')
+            if "linux" not in handler_config_json['handlerConfiguration']:
+                raise ExtensionHandlerConfigurationError('No linux configurations present in '
+                                                  'HandlerConfiguration')
+
+            try:
+                self.resource_config = ExtensionResourcesConfiguration(self.get_linux_configurations())
+            except ExtensionHandlerConfigurationError as e:
+                raise e
+
+            if self.resource_config or CGROUPS_ENFORCE_DEFAULT_LIMITS:
+                self.resource_limits = CGroupsLimits(self._name, self.resource_config)
+
+        except ExtensionHandlerConfigurationError as e:
+            logger.warn(ustr(e))
+            self.send_handler_configuration_event(message=ustr(e), is_success=False, log_event=False,
+                                                                  operation=WALAEventOperation.HandlerConfiguration)
+
+    def get_name(self):
+        return self._data.get("name", None)
+
+    def get_version(self):
+        return self._data.get("version", None)
+
+    def get_linux_configurations(self):
+        return self._data['handlerConfiguration']['linux']
```

Comment:
will throw if 'linux' is missing or has typo

Reply:
similar comment in other places (e.g. ExtensionResourcesConfiguration); we use the [] operator without previous validation that the key exists

Reply:
It was already been checked in the constructor. I'll change to look in `_data` rather than `handler_config_json`.

```
if "linux" not in self._data['handlerConfiguration']:
    raise ExtensionHandlerConfigurationError('No linux configurations present in '
                                             'HandlerConfiguration')
```